src/ddmd/cond.d

```diff
@@ -64,6 +67,227 @@ extern (C++) abstract class Condition : RootObject
+            loc,
+            aggrfe?cast(ForeachStatement)aggrfe.syntaxCopy():null,
+            rangefe?cast(ForeachRangeStatement)rangefe.syntaxCopy():null
+        );
+    }
+
+    private extern(D) void lowerArrayAggregate(Scope* sc)
+    {
+        auto aggr = aggrfe.aggr;
+        Expression el = new ArrayLengthExp(aggr.loc, aggr);
+        sc = sc.startCTFE();
+        el = el.semantic(sc);
+        sc = sc.endCTFE();
+        el = el.optimize(WANTvalue);
+        el = el.ctfeInterpret();
+        if (el.op == TOKint64){
+            dinteger_t length = el.toInteger();
+            auto es = new Expressions();
+            foreach(i;0..length)
+            {
+                auto index = new IntegerExp(loc, i, Type.tsize_t);
+                auto value = new IndexExp(aggr.loc, aggr, index);
+                es.push(value);
+            }
+            aggrfe.aggr = new TupleExp(aggr.loc, es);
+            aggrfe.aggr = aggrfe.aggr.semantic(sc);
+            aggrfe.aggr = aggrfe.aggr.optimize(WANTvalue);
+        }
+        else
+        {
+            aggrfe.aggr = new ErrorExp();
+            return;
+        }
+    }
+
+    private extern(D) Expression evaluate(Loc loc,Statement s)
+    {
+        auto tf = new TypeFunction(new Parameters(), null, 0, LINK.def, 0);
+        auto fd = new FuncLiteralDeclaration(loc, loc, tf, TOKreserved, null);
+        fd.fbody = s;
+        auto fe = new FuncExp(loc, fd);
+        auto ce = new CallExp(loc,fe,new Expressions());
+        return ce;
+    }
+
+    private extern(D) Statement createForeach(Loc loc, Parameters* parameters, Statement s)
+    {
+        if (aggrfe)
+        {
+            return new ForeachStatement(loc, aggrfe.op, parameters, aggrfe.aggr.syntaxCopy(), s, loc);
+        }
+        else
+        {
+            assert(!!rangefe && parameters.dim == 1);
+            return new ForeachRangeStatement(loc, rangefe.op, (*parameters)[0], rangefe.lwr.syntaxCopy(), rangefe.upr.syntaxCopy(), s, loc);
+        }
+    }
+
+    private extern(D) TypeStruct createTupleType(Loc loc, Expressions* e, Scope* sc)
+    {   // TODO: move to druntime?
+        auto sid = Identifier.generateId("Tuple");
+        auto sdecl = new StructDeclaration(loc, sid);
+        sdecl.storage_class |= STCstatic;
+        sdecl.members = new Dsymbols();
+        auto fid = Identifier.idPool(StaticForeach_tupleFieldName.ptr, StaticForeach_tupleFieldName.length);
+        auto ty = new TypeTypeof(loc, new TupleExp(loc, e));
+        sdecl.members.push(new VarDeclaration(loc, ty, fid, null, 0));
+        auto r = cast(TypeStruct)sdecl.type;
+        r.vtinfo = TypeInfoStructDeclaration.create(r); // prevent typeinfo from going to object file
+        return r;
+    }
+
+    private extern(D) Expression createTuple(Loc loc, TypeStruct type, Expressions* e)
+    {   // TODO: move to druntime?
+        return new CallExp(loc, new TypeExp(loc, type), e);
+    }
+
+    private void lowerNonArrayAggregate(Scope* sc)
```

Comment:
I know this stuff seems obvious to you (the author). The same thing happens to me. Then I come back to it 5 years later and wonder what the heck it is doing. The point of these reviews is to catch these sorts of things. It also helps a lot to see the comment which says "This does X" and then to look at the code, and see that indeed it does "X", and feel confident that you understand that section of code.